### ISSUE TITLE
[FEATURE] Improve getfacl/setfacl command | [mod=facl] [v0.5.5]

### ISSUE THREAD

**Issue by CkovMk** [x-cmd/x-cmd#138]
**Describe the feature**
ACL adds more control to filesystem permission on top of UGO permission. Traditionally its' controled by `getfacl` and `setfacl` commands, which is not intuitive to use. It would be great if x-cmd could implement an pretty TUI to visualize the acl rules.

**Is is urgent**
No. `getfacl` and `setfacl` works fine. However there's no other tools that enhances ACL management.

**Additional context**
None, just an feature request.

**Comment by edwinjhlee** [comment]
Thanks! I'm a bit embarrassed to admit I haven't actually used those two commands yet. I'll make some time to look into them.

---

After reviewing the `getfacl(1)` man page (https://www.mankier.com/1/getfacl), I believe this is feasible and will attempt an implementation in the latest versions. Thank you.

**Comment by curry798** [comment]
We've added the `facl` module in v0.5.5, which provides improved visualization of ACL rules.
You can use `x upgrade` to upgrade your x-cmd version and use it.

**Comment by edwinjhlee** [comment]
I am closing this issue. @CkovMk 
If you encouter problems, feel free to reopen it. Thank you.

**Comment by edwinjhlee** [comment]
Could you provide us some screenshots about ACL control like TureNAS's ACL webui.

For now, I won't be creating highly specialized tools, especially ones I don't know well. But I can keep an eye on this field.

If you know of any Rust or Go TUI tools for managing ACLs, please let us know and open a package recommendation issue.

**Comment by CkovMk** [comment]
> If you know of any Rust or Go TUI tools for managing ACLs, please let us know and open a package recommendation issue.

I ended up submitting a feature request here because I couldn't find any related TUI wrapper.

I tried TrueNAS long ago and not been using it for while. Its ACL management webui is impressive to me. Will come back with some screenshots once I got time to install it in a VM. And if you wish I could summarize a brief feature list as a starting point.

**Comment by edwinjhlee** [comment]
> > If you know of any Rust or Go TUI tools for managing ACLs, please let us know and open a package recommendation issue.
> 
> I ended up submitting a feature request here because I couldn't find any related TUI wrapper.
> 
> I tried TrueNAS long ago and not been using it for while. Its ACL management webui is impressive to me. Will come back with some screenshots once I got time to install it in a VM. And if you wish I could summarize a brief feature list as a starting point.

Thank you.

Here, I'd like to share the design philosophy behind x-cmd:
    
    1.  x-cmd is a shell library. The use cases for our TUI-enhanced tools are designed for common, but network-constrained and GUI-less, extreme operational environments.
    2.  For some specialized tools, the best app implementation might be in Rust/Go/Zig, or they should be designed as GUIs -- for example, webapp-based GUIs.
    3.  Therefore, for these types of tools, x-cmd essentially acts as a launcher. For example, using `x pkg` to quickly deploy the TUI app, or quickly deploy and launch a targeted GUI application.
    
Regarding the `acl` tool, I've found that the implementation differences between similar concepts on macOS and Linux are significant. Therefore, we can only have multiple solutions tailored to specific systems.
    
The current `facl` module in x-cmd focuses on collecting information to provide context for LLM or bug report communication.
    
To implement better ACL manipulation, it should be a more specialized file ACL management software, or developed on something like yazi superfile.
    
 I welcome you to provide more insights into the feature. We can first organize them and explore using similar functionality in more suitable open-source projects.
    
Thank you very much. I've benefited greatly from this discussion.

**Comment by CkovMk** [comment]
@edwinjhlee 

I agree it's not a simple task to build comprehensive  TUI for ACLs (Given how complicated TrueNAS has designed their implementation). And handling differences between OSes (like you mentioned, MacOS) adds another layer of difficulty. I do write some janky shell / python scripts to make my life easier dealing with ACLs, but would definitely like a more elegant solution.

I just freshly installed TrueNAS Scale 24.10 and brings some screenshots of its ACL interface. It turns out they redesigned it across the years, and it got even better.

Here's traditional UGO permission page (which is nothing special):

![Image](https://github.com/user-attachments/assets/ad5b6a4a-4ac4-4d55-ac2d-469c76233694)

Press the "ACL" button and select "Create a custom ACL", and we got this:

![Image](https://github.com/user-attachments/assets/79eed44e-eb16-4da9-8d11-ec952127b766)

- Owner and Owner Group can be edited at the top.
- Existing ACL entries are listed at bottom left. New ACL entry can be crerated, existing entry can be removed (except for default entries).
- Checkbox to select whether to apply the settings recursively.
- By switching between ACL entries, entry detail is ready for editing at the bottom right.
- When a new ACL entry is created, the ACL type must be selected:

![Image](https://github.com/user-attachments/assets/0648f6c0-80e1-4bf7-94c2-2e34952b5ad1)

- If ACL Type is "User" or "Group", the user / group name prompt shows up:

![Image](https://github.com/user-attachments/assets/4d525999-5653-405f-b2cd-addd5a9d4392)

- The RWX permission can be controlled via checkboxes, and an additional checkbox is used to control whether this item should be default (applies to newly-created files and directories).

It also includes ablilty to save and load ACL presets.

Here's an overview to all types of entries:

![Image](https://github.com/user-attachments/assets/8d644be7-6613-4cef-8236-047a81300976)
